src/utils/mod.rs

```diff
@@ -1010,3 +1012,154 @@ pub fn create_unrestricted_signing_rsa_public(
         .with_parms(PublicParmsUnion::RsaDetail(rsa_parms))
         .build() // should not fail as we control the params
 }
+
+// Enum with the bit flag for each PCR slot.
+#[derive(BitFlags, Hash, Debug, PartialEq, Clone, Copy)]
+#[repr(u32)]
+pub enum PcrSlot {
+    Slot0 = 0x0000_0001,
+    Slot1 = 0x0000_0002,
+    Slot2 = 0x0000_0004,
+    Slot3 = 0x0000_0008,
+    Slot4 = 0x0000_0010,
+    Slot5 = 0x0000_0020,
+    Slot6 = 0x0000_0040,
+    Slot7 = 0x0000_0080,
+    Slot8 = 0x0000_0100,
+    Slot9 = 0x0000_0200,
+    Slot10 = 0x0000_0400,
+    Slot11 = 0x0000_0800,
+    Slot12 = 0x0000_1000,
+    Slot13 = 0x0000_2000,
+    Slot14 = 0x0000_4000,
+    Slot15 = 0x0000_8000,
+    Slot16 = 0x0001_0000,
+    Slot17 = 0x0002_0000,
+    Slot18 = 0x0004_0000,
+    Slot19 = 0x0008_0000,
+    Slot20 = 0x0010_0000,
+    Slot21 = 0x0020_0000,
+    Slot22 = 0x0040_0000,
+    Slot23 = 0x0080_0000,
+}
+
+/// A struct representing pcr selections
+///
+/// The minimum number of octets allowed in a TPMS_PCR_SELECT.sizeOfSelect
+/// is not determined by the number of PCR implemented but by the
+/// number of PCR required by the platform-specific
+/// specification with which the TPM is compliant or by the implementer if
+/// not adhering to a platform-specific specification.
+#[derive(Debug, Default, Clone)]
+pub struct PcrSelections {
+    size_of_select: u8,
+    items: HashMap<HashingAlgorithm, BitFlags<PcrSlot>>,
+}
+
+impl From<PcrSelections> for TPML_PCR_SELECTION {
+    fn from(pcr_selections: PcrSelections) -> TPML_PCR_SELECTION {
+        let mut ret: TPML_PCR_SELECTION = Default::default();
+        for (hash_algorithm, pcr_slots) in pcr_selections.items {
+            let tpms_pcr_selection = &mut ret.pcrSelections[ret.count as usize];
+            tpms_pcr_selection.hash = hash_algorithm.clone().into();
+            tpms_pcr_selection.sizeofSelect = pcr_selections.size_of_select;
+            tpms_pcr_selection.pcrSelect = pcr_slots.bits().to_le_bytes();
+            ret.count += 1;
+        }
+        ret
+    }
+}
+
+impl TryFrom<TPML_PCR_SELECTION> for PcrSelections {
+    type Error = Error;
+    fn try_from(tpml_pcr_selection: TPML_PCR_SELECTION) -> Result<PcrSelections> {
+        let mut ret: PcrSelections = Default::default();
+        let mut size_of_select: Option<u8> = None;
+        // Loop over available selections
+        for selection_index in 0..(tpml_pcr_selection.count as usize) {
+            let selection = &tpml_pcr_selection.pcrSelections[selection_index];
+            let pcr_slots: BitFlags<PcrSlot> =
+                BitFlags::<PcrSlot>::try_from(u32::from_le_bytes(selection.pcrSelect)).unwrap();
```

Comment:
We try to not use panicking function in our code. Is it possible to remove the `unwrap` with something like:
```rust
.or_else(|e| {
    error!("Error converting pcrSelect to a BitFlags<PcrSlot>: {}.", e);
    Err(Error::local_error(WrapperErrorKind::InvalidParam))
})?;
```
And similar kind of things for all the `unwrap`. Ideally we always want to return an error instead of panicking.

Reply:
Sorry did not mention and I think it is not imported in that file, but `error!` is imported with `use log::error;`

Reply:
You are right I was thinking that it could not fail. Because I only made it possible to select 24 slots through the enum. But it is also possible to send in something that does not come from the tpm and try to convert it and in that case that would panic. I use your code and make it more robust against conversion of invalid data.